File: .tmp_py4013/batch_04_sources/2519_count_the_number_of_k_big_indices.js
// LeetCode 2519 - Count the Number of K-Big Indices
// https://leetcode.com/problems/count-the-number-of-k-big-indices/

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var kBigIndices = function(nums, k) {
    const Fenwick = function(n) {
        this.bit = Array(n + 2).fill(0);
    };
    Fenwick.prototype.add = function(i, v) {
        for (; i < this.bit.length; i += i & -i) this.bit[i] += v;
    };
    Fenwick.prototype.sum = function(i) {
        let s = 0;
        for (; i > 0; i -= i & -i) s += this.bit[i];
        return s;
    };
    const n = nums.length;
    const uniq = nums.slice().sort((a, b) => a - b);
    let m = 0;
    for (let i = 0; i < uniq.length; i++) {
        if (i === 0 || uniq[i] !== uniq[i - 1]) uniq[m++] = uniq[i];
    }
    const rank = new Map();
    for (let i = 0; i < m; i++) rank.set(uniq[i], i + 1);
    const left = Array(n), right = Array(n);
    let ft = new Fenwick(m);
    for (let i = 0; i < n; i++) {
        const r = rank.get(nums[i]);
        left[i] = ft.sum(r - 1);
        ft.add(r, 1);
    }
    ft = new Fenwick(m);
    for (let i = n - 1; i >= 0; i--) {
        const r = rank.get(nums[i]);
        right[i] = ft.sum(r - 1);
        ft.add(r, 1);
    }
    let ans = 0;
    for (let i = 0; i < n; i++) {
        if (left[i] >= k && right[i] >= k) ans++;
    }
    return ans;
};

File: .tmp_rs4013/rust_check09/2519_count_the_number_of_k_big_indices.rs
struct Solution;

// LeetCode 2519 - Count the Number of K-Big Indices
// https://leetcode.com/problems/count-the-number-of-k-big-indices/

use std::collections::HashMap;

struct Fenwick {
    bit: Vec<i32>,
}

impl Fenwick {
    fn new(n: usize) -> Self {
        Self { bit: vec![0; n + 2] }
    }
    fn add(&mut self, mut i: usize, v: i32) {
        while i < self.bit.len() {
            self.bit[i] += v;
            i += i & i.wrapping_neg();
        }
    }
    fn sum(&self, mut i: usize) -> i32 {
        let mut s = 0;
        while i > 0 {
            s += self.bit[i];
            i -= i & i.wrapping_neg();
        }
        s
    }
}

impl Solution {
    pub fn k_big_indices(nums: Vec<i32>, k: i32) -> i32 {
        let n = nums.len();
        let mut uniq = nums.clone();
        uniq.sort_unstable();
        uniq.dedup();
        let mut rank = HashMap::new();
        for (i, &v) in uniq.iter().enumerate() {
            rank.insert(v, i + 1);
        }
        let m = uniq.len();
        let mut left = vec![0; n];
        let mut right = vec![0; n];
        let mut ft = Fenwick::new(m);
        for i in 0..n {
            let r = rank[&nums[i]];
            left[i] = ft.sum(r - 1);
            ft.add(r, 1);
        }
        let mut ft = Fenwick::new(m);
        for i in (0..n).rev() {
            let r = rank[&nums[i]];
            right[i] = ft.sum(r - 1);
            ft.add(r, 1);
        }
        let mut ans = 0;
        for i in 0..n {
            if left[i] >= k && right[i] >= k {
                ans += 1;
            }
        }
        ans
    }
}

fn main() {}
